### ISSUE TITLE
Need help -- Have you sorted boundry boxes?

### ISSUE THREAD

**Issue by nbhupendra** [clovaai/CRAFT-pytorch#52]
Boundry boxes are shuffled. Can you please help me for this. Really appreciate.

**Comment by hanish3464** [comment]
What data are you talking about

**Comment by nbhupendra** [comment]
i am talking about detected text bounding_box in image. I have got the bounding box coordinates as txt files. The resulted bounding boxes are shuffled and i could not sort it . When there is a presence of curved text in a same line like in the image below, the order gets shuffled and i need to sort it before passing it to text extraction model.

**Comment by nbhupendra** [comment]
Thanks for suggestion however i want it line by line as they appear so i am not sure sorting from y value with work. Can you please share some sample if possible.
Thanks

**Comment by hanish3464** [comment]
Could you please share difficult sample images to sort??? I would like to think about this problem. I'll share it if I find a good solution!

**Comment by nbhupendra** [comment]
Ok you can refer https://github.com/clovaai/CRAFT-pytorch/issues/43#issue-491481824 
its same problem and image also can be utilize same for solutioning.

**Comment by YacobBY** [comment]
@nbhupendra The boundary boxes are sorted by height of beginning (top to bottom). The highest textboxes are  shown first. You probably want to sort them from left to right

**Comment by nbhupendra** [comment]
> @nbhupendra The boundary boxes are sorted by height of beginning (top to bottom). The highest textboxes are shown first. You probably want to sort them from left to right

Yes i want them to sort them from left to right line by line. @YacobBY can you please help me for this.

**Comment by YacobBY** [comment]
@nbhupendra Yeah sure something like this should work. 
```
def findBoxes(image, net):
    # load data
    # print("Craft image shape:", image.shape) #IMAGE SIZE AND WIDTH
    bboxes, polys, score_text = test_net(net, image, args.text_threshold, args.link_threshold, args.low_text, args.cuda, poly=True)

    boxPositions = []
    for box in (bboxes):
        # Get coordtinates for bounding boxes here
        boxPolys = np.array(box).astype(np.int32).reshape((-1)) 
        boxPolys[boxPolys < 0] = 0  # Negative values will be set to zero
        # 
        minY, maxY, minX, maxX = boxPolys[1], boxPolys[5], boxPolys[0], boxPolys[2]
        # print("bbox array:", poly)
        boxPositions.append((minY, maxY, minX, maxX))

    boxPositions = sorted(boxPositions,key=lambda x: x[3]) # sorts boxes by 2rd X index, most leftern X values first.
    # Extract the boxpositions as images
    croppedTextBoxImages = []
    for pos in (boxPositions):
        croppedTextBoxImages.append(image[pos[0]:pos[1], pos[2]: pos[3]])
    return image, croppedTextBoxImages, boxPositions

```

**Comment by nbhupendra** [comment]
> @nbhupendra Yeah sure something like this should work. Didn't test this with that new "refine net" thing though you might need to add that parameter to the test_net function to make it work
> 
> ```
> def findBoxes(image, net):
>     bboxes, polys, score_text = test_net(net, image, args.text_threshold, args.link_threshold, args.low_text, args.cuda, poly=True)
> 
>     boxPositions = []
>     for box in (bboxes):
>         # Get coordtinates for bounding boxes here
>         boxPolys = np.array(box).astype(np.int32).reshape((-1)) 
>         boxPolys[boxPolys < 0] = 0  # Negative values will be set to zero
>         # 
>         minY, maxY, minX, maxX = boxPolys[1], boxPolys[5], boxPolys[0], boxPolys[2]
>         # print("bbox array:", poly)
>         boxPositions.append((minY, maxY, minX, maxX))
> 
>     boxPositions = sorted(boxPositions,key=lambda x: x[3]) # sorts boxes by 2rd X index, most leftern X values first.
>     # Extract the boxpositions as images
>     croppedTextBoxImages = []
>     for pos in (boxPositions):
>         croppedTextBoxImages.append(image[pos[0]:pos[1], pos[2]: pos[3]])
>     return image, croppedTextBoxImages, boxPositions
> ```

Thank you. I tried it but seems it does not full filled  my requirement. Let me explain, may be then you can help me better.

![test](https://user-images.githubusercontent.com/49706661/66050045-1722ad80-e54a-11e9-8c83-4c2769f9593d.png)

For the above image, the model outputs a txt file in which the bb coordinates are as follows. I have added the detected text for better explanation of my problem. In this case the detection order is

146,36,354,34,354,82,146,84 "Australian"

273,78,434,151,411,201,250,129 "Collection"

146,97,250,97,250,150,146,150 "vine"

77,166,131,126,154,158,99,197 "Old"

242,215,361,241,354,273,235,248 "Valley"

140,247,224,219,234,250,150,277 "Eden"

194,298,306,296,307,324,194,325 "Shiraz"

232,406,363,402,364,421,233,426 "Vintage"

152,402,216,405,215,425,151,422 "2008"

124,470,209,480,207,500,122,490 "South"

227,481,387,472,389,494,228,503 "Australia"

222,562,312,564,311,585,222,583 "Gibson"

198,564,217,564,217,584,198,584 "by"

386,570,421,570,421,600,386,600 "750 ml"

But the expected output is [Australian->old->vine->collection->Eden->Valley->shiraz->2008->vintage->south->Australia->by->GIBSON->750ml.]

So basically i need from top left first how it was read. Please see if you could make it for me. Really appreciate for your help.

**Comment by YacobBY** [comment]
Ehhh your case is actually rather difficult but I think there's still a solution (as long as you're not trying to look for text diagonally)

First you make an array of all the boxes, then you sort that array by the most leftern boxes.
You then take both the highest and lowest point (Y coordinates) of the first text box in the array. Then you look through the entire array and see if there are any boundary boxes in the rest of the array which have a Y coordinate that is between the highest and lowest point of those boundary boxes. You then add the most leftern word along with any other boundary boxes that are in the same height to an array of sentences. Then make an array of those sentences (array of arrays) and sort them by the height of their first word (index) which will give you all the sentences in order from top to bottom as well. Then you have to select each word in the sentence list and put them all back in one array. 

A hand drawn example: all the words after the word Krishna fall in the same line.
![66052164-c44af500-e54d-11e9-8f53-09bb94a063ce(1)](https://user-images.githubusercontent.com/30029456/66054018-aa041d80-e533-11e9-993a-90209016923c.png)

**Comment by YacobBY** [comment]
Don't have the time for that currently, the guidelines I wrote out shouldn't be that hard to implement though, just going to cost a bit of time finding out how to sort the boxes after each other by height.

**Comment by nbhupendra** [comment]
> @nbhupendra Yeah sure something like this should work. Didn't test this with that new "refine net" thing though you might need to add that parameter to the test_net function to make it work
> 
> ```
> def findBoxes(image, net):
>     bboxes, polys, score_text = test_net(net, image, args.text_threshold, args.link_threshold, args.low_text, args.cuda, poly=True)
> 
>     boxPositions = []
>     for box in (bboxes):
>         # Get coordtinates for bounding boxes here
>         boxPolys = np.array(box).astype(np.int32).reshape((-1)) 
>         boxPolys[boxPolys < 0] = 0  # Negative values will be set to zero
>         # 
>         minY, maxY, minX, maxX = boxPolys[1], boxPolys[5], boxPolys[0], boxPolys[2]
>         # print("bbox array:", poly)
>         boxPositions.append((minY, maxY, minX, maxX))
> 
>     boxPositions = sorted(boxPositions,key=lambda x: x[3]) # sorts boxes by 2rd X index, most leftern X values first.
>     # Extract the boxpositions as images
>     croppedTextBoxImages = []
>     for pos in (boxPositions):
>         croppedTextBoxImages.append(image[pos[0]:pos[1], pos[2]: pos[3]])
>     return image, croppedTextBoxImages, boxPositions
> ```
i guess this suppose to be - boxPolys[1], boxPolys[5], boxPolys[0], boxPolys[2] to boxPolys[1], boxPolys[3], boxPolys[0], boxPolys[2] right?

**Comment by YacobBY** [comment]
@nbhupendra The Y axis in OpenCV is pretty stupid. it is from top to bottom instead of bottom to top. The poly positions should stay consistent if this repo didn't change any of the annotation positions since I wrote this code

So first sort by x (boxPositions,key=lambda x: x[3]) for left to right and then check for each index if any index below it it is in range for their minY to maxY. Then take out all  those found boxes and put them in a separate array. do it again until the original array is empty. Then sort those separate arrays by their minY positions and you should have "sentences" from top to bottom.

**Comment by nbhupendra** [comment]
@YacobBY i tried below code that works in some of cases but not in all and also not consistent. can you please check if it can be improve.

nearest = max_height * .7

boxPositions = sorted(boxPositions, key=lambda r: (int(nearest* round(float(r[0]) / nearest)) * max_width + r[2]))

**Comment by hbellafkir** [comment]
@nbhupendra did you found a solution?

**Comment by programmer8396** [comment]
Hy Everyone 
https://github.com/facebookresearch/detectron2/issues/2420

I have the same issue for sorting the bounding box from left to right and top to bottom 
Please post if you have any suggestion.

**Comment by Batterlake** [comment]
In case someone stumbles upon this issue, there are number of implementations one can use from one of the popular OCR frameworks:
- get_paragraph in [EasyOCR](https://github.com/JaidedAI/EasyOCR/blob/master/easyocr/utils.py#L643)
- sorted_layout_boxes in [paddlestructure](https://github.com/PaddlePaddle/PaddleOCR/blob/0850586667308d38e113447e8d095e955092fe53/ppstructure/recovery/recovery_to_doc.py#L81)
- sort_boxes in [paddleocr](https://github.com/PaddlePaddle/PaddleOCR/blob/release/2.7/tools/infer/predict_system.py#L123)
These methods are not perfect and do not account for angled/curved text.

There are number of frameworks that approach this problem in seq2seq-manner, for example: [LayoutReader](https://github.com/microsoft/unilm/tree/master/layoutreader).
